package zurg;

option cc_generic_services = true;
option java_generic_services = true;
option py_generic_services = true;
option java_package = "com.chenshuo.muduo.zurg.proto";
option java_outer_classname = "MasterProto";

import "muduo/protorpc2/rpc2.proto";
import "slave.proto";

message SlaveHeartbeat {
  required string slave_name = 1;
  required int64 send_time_us = 2;

  // static data
  optional string host_name = 20;
  optional int32 listen_port = 21;
  optional int32 slave_pid = 22;
  optional int64 start_time_us = 23;
  optional string slave_version = 24;
  repeated string env_vars = 25;

  // static proc files per hour
  optional string cpuinfo = 30;
  optional string version = 31;
  optional string etc_mtab = 32;
  optional string sysctl = 33;
  message Uname {
    required string sysname = 1;
    required string nodename = 2;
    required string release = 3;
    required string version = 4;
    required string machine = 5;
    required string domainname = 6;
  }
  optional Uname uname = 34;

  // dynamic proc files 10s
  optional string meminfo = 40;
  optional string proc_stat = 41;
  optional string loadavg = 42;
  optional string diskstats = 43;
  optional string net_dev = 44;
  optional string net_tcp = 45;

  // not so dynamic per minute
  repeated DiskUsage disk_usage = 50;
}

service MasterService {
  rpc slaveHeartbeat (SlaveHeartbeat) returns (rpc2.Empty) {
    option (rpc2.noreturn) = true;
  }
  rpc appStatusChange (ApplicationStatus) returns (rpc2.Empty) {
    option (rpc2.noreturn) = true;
  }
}

